VSS.md

```diff
@@ -0,0 +1,518 @@
+
+- Requirement: Index creation triggers a backfill process which indexes the existing keyspace entries.
+
+**Hybrid Query CUJ** \- As a user, I want to be able to perform hybrid queries to incorporate filtering criteria in addition to best-match vector search
+
+- Requirement: Support for filtering numeric and tag-based filters.  
+- Requirement: Support for boolean combinations of operators in combination with filtering (AND, OR, NOT)
+
+**Result Set Management CUJ** \- As a user I want to dictate which fields are returned in a search
+
+- Requirement: Support query controls to set the desired returned fields.
+
+**ACLs CUJ** \- As a user of Access Control Lists (ACLs), I want to manage user-based permissions, specific command permissions, etc to manage vector search access
+
+- Requirement: Existing ACls functionality should extend to vector search commands
+
+**Persistence CUJ** \- As a user I want to be able to leverage RDB and/or AOF for persisting and loading indexes
+
+- Requirement: Index metadata should be persisted leveraging RDB  
+- Requirement: Index types that take significant time to rebuild, such as vector indexes, must support persistence of the index payload.
+
+**Tuning CUJ** \- As a user who is also managing the infrastructure, I want to tune my configurations, like number of threads, to optimize my search performance
+
+- Requirement: Configs like number of threads should be exposed and tunable to users
+
+## Current Specification
+
+The search module allows a user to create any number of indexes. Each index consists of any number of typed fields and covers a specified subset of the keys within a Valkey database.
+
+The creation of an index initiates a backfill process in the background that will scan the keyspace and insert all covered keys into the newly created index. Mutations of a key are automatically reflected in the index. This process is automatic after the creation of the index. Mutation of keys is explicitly allowed during the backfill process. The current state of a backfill may be interrogated by a client using the FT.INFO command.
+
+A query operation is a two step process. In the first step, a query is executed that generates a set of keys and distances. In the second step, that list of keys and distances is used to generate a response.
+
+The first step is performed in a background thread, while the second step is performed by the mainthread. If a key produced by step 1 is mutated before step 2 is executed by the mainthread, the result of the query may not be consistent with the latest state of the keys.
+
+### Cluster Mode Operation
+
+In cluster mode, indexes are sharded just like data, meaning that each node maintains an index only for the keys that are present on that node.
+This means for that for data mutation operations, no cross-cluster communication is required, each node simply updates it's local index to reflect the mutation of the local key.
```

Comment:
Does this mean that indexes don't need to be replicated? Please elaborate and mention something about these cases:

* Are indexes replicated and stored in RDB and AOF files?
* Do the indexes not need to be replicated, because the replica builds its own index from the keys it receives from the primary?
* The same questions about slot migration: Do the indexes need to be moved with slot migration or does the importing node just build its own index with the keys it receives from the migrating node?